### PR TITLE
[Feature] Add unlock token mechanism

### PR DESCRIPTION
**Pull request by CaptainVincent**
**Unlock token mechanism is base on `Token Economics` in [Oasis ETH gas token](https://hackmd.io/@PIv_zaEuR2e5PtCzEc4HRA/HkbOgO47v)**

- Summary
There is a total fixed supply of 21 million OETHs. The blockchain unlocks 525,000 OETHs every month in the first 20 months after the mainnet launch. The monthly release is cut to 1/2 every 20 months. It takes perpetuity to release the entire 21 million OETHs.

- Implementation (update 2020.11.6)
  1. ~~We expect unlock token once every month when runtime get the first tx in the month.~~
  2. Use fixed 30 days as unit round. The release amount will cut to 1/2 every 20 rounds.
  3. Release amount in a single round is depending on account for what is the percentage in this round.
  4. ~~If runtime doesn't get any tx in expecting month,~~ The unhandled release quota will pending until getting any transaction. 
  5. We keep some information (initial timestamp, ~~unlock round, unlock token amount~~, pending round, unlocked ticks) in special account's storage for making every compute nodes reach a consensus.
  > - `initial timestamp` means the start time to apply unlock token mechanism.
  > - `pending round` means which round is not yet fully unlocked.
  > - `unlocked ticks` means already unlocked ticks/seconds in last unlocked round. 
  > - Below is more detail of the used terms in the implementation about calculating to unlock amount.
fig

- Testing
run `cargo test --features="test"`.
The result should be the same as the following content. 
```
(omitted)

running 8 tests
test fund::tests::test_try_unlock_1_round ... ok
test fund::tests::test_try_unlock_10_rounds ... ok
test fund::tests::test_try_unlock_sequential_3_ticks ... ok
test fund::tests::test_try_unlock_zero ... ok
test fund::tests::test_try_unlock_random_txs_20rounds ... ok
test fund::tests::test_try_unlock_random_txs_cross_20rounds ... ok
test fund::tests::test_try_unlock_100_rounds ... ok
test fund::tests::test_try_unlock_1000_rounds ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

(omitted)
```

### CODE REVIEW

File: src/fund.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+use oasis_ethwasi_runtime_common::parity::NullBackend;
+extern crate chrono;
+use chrono::{DateTime, NaiveDateTime, Utc, Datelike};
+extern crate date_time;
+use date_time::month_tuple::MonthTuple;
+use std::str::FromStr;
+use std::convert::TryInto;
+use byteorder::{BigEndian, ReadBytesExt};
+
+fn timestamp_to_mt(timestamp:i64) -> MonthTuple{
```
Comment (hydai) [inline src/fund.rs]: Should we run a fmt tool for checking the coding style?

File: src/fund.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    const FACTOR: f32 = 0.5;
+    /// Total Amount.
+    const TOTAL_AMOUNT: u32 = 21000000;
+    /// Initial unlock bucket.
+    const INITIAL_BUCKET: f32 = 525000.0;
+
+    pub fn try_unlock(timestamp:u64, state:&mut State<NullBackend>) {
+        let beneficiary = Address::from_str(FundManager::BENEFICIARY).unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(0)).unwrap();
+        let init_timestamp = value.get(24..).unwrap().read_i64::<BigEndian>().unwrap();
```
Comment (hydai) [inline src/fund.rs:45]: What does the `24` stand for?

File: src/fund.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        let value = state.storage_at(&beneficiary, &H256::from(1)).unwrap();
+        let mut unlock_round = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
```
Comment (hydai) [inline src/fund.rs]: What does the `28` stand for?

File: src/fund.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        let value = state.storage_at(&beneficiary, &H256::from(2)).unwrap();
+        let mut unlock_amount = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
```
Comment (hydai) [inline src/fund.rs]: What does the `28` stand for?

File: src/fund.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    /// Adjust period.
+    const PERIOD: u32 = 20;
+    // Factor
+    const FACTOR: f32 = 0.5;
+    /// Total Amount.
+    const TOTAL_AMOUNT: u32 = 21000000;
+    /// Initial unlock bucket.
+    const INITIAL_BUCKET: f32 = 525000.0;
+
+    pub fn try_unlock(timestamp:u64, state:&mut State<NullBackend>) {
```
Comment (hydai) [inline src/fund.rs]: Should we make this function atomic?
Once it failed, we should rollback states.

**Review by hydai — CHANGES_REQUESTED** [review]

File: src/fund.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        let mut unlock_round = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(2)).unwrap();
+        let mut unlock_amount = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+
+        if unlock_amount < FundManager::TOTAL_AMOUNT {
+            let mut last_ym = timestamp_to_mt(init_timestamp as i64);
+            last_ym.add_months(unlock_round);
+            let cur_ym = timestamp_to_mt(timestamp as i64);
+            let mut funding = 0;
+            while last_ym <= cur_ym && unlock_amount < FundManager::TOTAL_AMOUNT{
```
Comment (CaptainVincent) [inline src/fund.rs]: I think I should use `last_ym < cur_ym` here. I will fixup it.

File: src/fund.rs (hunk 3 above)
Comment (CaptainVincent) [reply src/fund.rs]: get byte slice from position 28 to 32 (last 4 bytes). use it and convert to u32.

File: src/fund.rs (hunk 4 above)
Comment (CaptainVincent) [reply src/fund.rs]: same as before comment.

File: src/fund.rs (hunk 2 above)
Comment (CaptainVincent) [reply src/fund.rs:45]: get byte slice from position 24 to 32 (last 8 bytes). use it and convert to i64.

Comment (hydai) [reply src/fund.rs:45]: We should add a comment to describe the data layout?

File: src/fund.rs (hunk 3 above)
Comment (hydai) [reply src/fund.rs]: Ditto.

File: src/fund.rs (hunk 4 above)
Comment (hydai) [reply src/fund.rs]: Ditto.

File: src/fund.rs (hunk 1 above)
Comment (CaptainVincent) [reply src/fund.rs]: Yes, i will accomplish it later.

File: src/fund.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        // |             2 |            reserved        |   amount  |
+        // +---------------+----------+-----------------+-----------+
+        let value = state.storage_at(&beneficiary, &H256::from(0)).unwrap();
+        let init_timestamp = value.get(24..).unwrap().read_i64::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(1)).unwrap();
+        let mut unlock_round = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(2)).unwrap();
+        let mut unlock_amount = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+
+        if unlock_amount < FundManager::TOTAL_AMOUNT {
```
Comment (hydai) [inline src/fund.rs]: We should also add comments to explain our unlock logic.

**Review by hydai — CHANGES_REQUESTED** [review]

File: src/fund.rs (hunk 5 above)
Comment (CaptainVincent) [reply src/fund.rs]: I think this case should be covered by oasis-parity state.commit function (I saw oasis-parity use storage_cache internal) and consensus layer.
If some compute node fail within execution then its' state change should not be acceptable by others and need drop & re-sync.
Do my opinion is reasonable?

Comment (hydai) [reply src/fund.rs]: It makes sense.

File: src/fund.rs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+impl FundManager {
+    /// Unlock address.
+    const BENEFICIARY: &'static str = "7110316b618d20d0c44728ac2a3d683536ea682b";
+    /// Adjust period.
+    const PERIOD: u32 = 20;
+    // Factor
+    const FACTOR: f32 = 0.5;
+    /// Total Amount.
+    const TOTAL_AMOUNT: u32 = 21000000;
```
Comment (dm4) [inline src/fund.rs]: Should we use `wei` as unit here?

Comment (dm4) [reply src/fund.rs]: After using `wei` as unit, we should use `u256` here.

File: src/fund.rs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    /// Initial unlock bucket.
+    const INITIAL_BUCKET: f32 = 525000.0;
```
Comment (dm4) [inline src/fund.rs]: Should we use `wei` as unit and use `u256` as type here?

File: src/fund.rs (hunk 8 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

File: src/fund.rs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+use ethereum_types::{Address, H256, U256};
+use oasis_ethwasi_runtime_common::parity::NullBackend;
+extern crate chrono;
+use chrono::{DateTime, Datelike, NaiveDateTime, Utc};
+extern crate date_time;
+use byteorder::{BigEndian, ReadBytesExt};
+use date_time::month_tuple::MonthTuple;
+use std::{convert::TryInto, str::FromStr};
+
+fn timestamp_to_mt(timestamp: i64) -> MonthTuple {
```
Comment (hydai) [inline src/fund.rs]: We need more comments to explain this function.

File: src/fund.rs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        let mut unlock_amount: U256 = value.into();
+
+        if unlock_amount < total_amount {
+            let mut last_ym = timestamp_to_mt(init_timestamp as i64);
+            last_ym.add_months(unlock_round);
+            let cur_ym = timestamp_to_mt(timestamp as i64);
+            let mut funding = U256::from(0);
+            let mut exponent = 0;
+            let mut bucket = initial_bucket;
+            while last_ym < cur_ym && unlock_amount < total_amount {
```
Comment (hydai) [inline src/fund.rs]: I believe there should be comments to explain the details of this while loop.

File: src/fund.rs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        // +---------------+----------+-----------------+-----------+
+        let value = state.storage_at(&beneficiary, &H256::from(0)).unwrap();
+        let init_timestamp = value.get(24..).unwrap().read_i64::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(1)).unwrap();
+        let mut unlock_round = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(2)).unwrap();
+        let mut unlock_amount: U256 = value.into();
+
+        if unlock_amount < total_amount {
+            let mut last_ym = timestamp_to_mt(init_timestamp as i64);
```
Comment (hydai) [inline src/fund.rs]: You variable name should have more comments.

**Review by hydai — DISMISSED** [review]
Please add more comments. Thanks.

File: src/fund.rs (hunk 11 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

File: src/fund.rs (hunk 10 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

File: src/fund.rs (hunk 12 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

File: src/fund.rs [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    // Factor
+    const FACTOR: u32 = 2;
+    /// Total Amount is 21000000000000000000000000 wei.
+    const TOTAL_AMOUNT: &'static str = "115EEC47F6CF7E35000000";
+    /// Initial unlock bucket is 525000000000000000000000 wei.
+    const INITIAL_BUCKET: &'static str = "6F2C4E995EC98E200000";
+
+    // There is a total fixed supply of 21 million OETHs.
+    // The blockchain unlocks 525,000 OETHs every month in the first 20 months and
+    // the monthly release is cut to 1/2 every 20 months.
```
Comment (dm4) [inline src/fund.rs]: Should we move these comment before variable declaration?

File: src/fund.rs [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+pub struct FundManager;
+
+impl FundManager {
+    /// Unlock address.
+    const BENEFICIARY: &'static str = "7110316b618d20d0c44728ac2a3d683536ea682b";
+    /// Adjust period.
+    const PERIOD: u32 = 20;
+    // Factor
```
Comment (dm4) [inline src/fund.rs]: Why only factor use `//` comment? XD

File: src/fund.rs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            let mut bucket = initial_bucket;
+
+            // if last handled (year, month) still before current (year, month) and
+            // still have funding not yet unlocked.
+            while last_ym < cur_ym && unlock_amount < total_amount {
+                last_ym.add_months(1);
+                unlock_round += 1;
+                if exponent != unlock_round / FundManager::PERIOD {
+                    exponent = unlock_round / FundManager::PERIOD;
+                    bucket = initial_bucket / U256::from(FundManager::FACTOR.pow(exponent));
```
Comment (dm4) [inline src/fund.rs]: `FACTOR.pow(exponent)` will overflow if `exponent > 31`

File: src/fund.rs (hunk 14 above)
Comment (CaptainVincent) [reply src/fund.rs]: my mistake.

File: src/fund.rs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+            let cur_ym = timestamp_to_mt(timestamp as i64);
+            let mut funding = U256::from(0);
+            let mut exponent = 0;
+            let mut bucket = initial_bucket;
+
+            // if last handled (year, month) still before current (year, month) and
+            // still have funding not yet unlocked.
+            while last_ym < cur_ym && unlock_amount < total_amount {
+                last_ym.add_months(1);
+                unlock_round += 1;
```
Comment (dm4) [inline src/fund.rs]: If we update `unlock_round` here, we'll miss 1 round and unlock 525,000 OETH only for 19 rounds.

File: src/fund.rs (hunk 15 above)
Comment (CaptainVincent) [reply src/fund.rs]: Yes, you are right.
Do we need consider this scenario?
This case means (20 months)*31 / 12 months almost equal 52 years.
If we try to cover all cases, at least need 10 bytes integer.
(525000000000000000000000 wei = 0x6F2C4E995EC98E200000)

Comment (hydai) [reply src/fund.rs]: We should avoid any overflow issue even if it may happen in the future.

File: src/fund.rs (hunk 14 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

File: src/fund.rs (hunk 13 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

File: src/fund.rs (hunk 15 above)
Comment (CaptainVincent) [reply src/fund.rs]: done. I use type U256 handle it finally.

File: src/fund.rs (hunk 16 above)
Comment (CaptainVincent) [reply src/fund.rs]: done.

**Comment by hydai** [comment]
@CaptainVincent Could you add a test to cover this feature? Using `#[test]`, so we can run tests with `cargo test`.

**Review by dm4 — APPROVED** [review]

**Comment by CaptainVincent** [comment]
> @CaptainVincent Could you add a test to cover this feature? Using `#[test]`, so we can run tests with `cargo test`.

done.

**Review by hydai — APPROVED** [review]
LGTM

**Review by dm4 — APPROVED** [review]
Nice pic! LGTM!

**Review by hydai — APPROVED** [review]
Let's launch the testnet for unlocking mechanism next week. Well done! 👍

**Pull request merged**